pkg/controller/disruption/disruption.go

```diff
@@ -0,0 +1,587 @@
+	glog.Infof("update DB %q", pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) removeDb(obj interface{}) {
+	pdb := obj.(*policy.PodDisruptionBudget)
+	glog.Infof("remove DB %q", pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) addPod(obj interface{}) {
+	pod := obj.(*api.Pod)
+	glog.Infof("addPod called on pod %q", pod.Name)
+	pdb := dc.getPdbForPod(pod)
+	if pdb == nil {
+		glog.Infof("No matching pdb for pod %q", pod.Name)
+		return
+	}
+	glog.Infof("addPod %q -> PDB %q", pod.Name, pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) updatePod(old, cur interface{}) {
+	pod := cur.(*api.Pod)
+	glog.Infof("updatePod called on pod %q", pod.Name)
+	pdb := dc.getPdbForPod(pod)
+	if pdb == nil {
+		glog.Infof("No matching pdb for pod %q", pod.Name)
+		return
+	}
+	glog.Infof("updatePod %q -> PDB %q", pod.Name, pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) deletePod(obj interface{}) {
+	pod := obj.(*api.Pod)
+	glog.Infof("deletePod called on pod %q", pod.Name)
+	pdb := dc.getPdbForPod(pod)
+	if pdb == nil {
+		glog.Infof("No matching pdb for pod %q", pod.Name)
+		return
+	}
+	glog.Infof("deletePod %q -> PDB %q", pod.Name, pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) enqueuePdb(pdb *policy.PodDisruptionBudget) {
+	key, err := controller.KeyFunc(pdb)
+	if err != nil {
+		glog.Errorf("Cound't get key for PodDisruptionBudget object %+v: %v", pdb, err)
+		return
+	}
+	dc.queue.Add(key)
+}
+
+func (dc *DisruptionController) getPdbForPod(pod *api.Pod) *policy.PodDisruptionBudget {
+	// GetPodPodDisruptionBudgets returns an error only if no
+	// PodDisruptionBudgets are found.  We don't return that as an error to the
+	// caller.
+	pdbs, err := dc.pdbLister.GetPodPodDisruptionBudgets(pod)
+	if err != nil {
+		glog.Infof("No PodDisruptionBudgets found for pod %v, PodDisruptionBudget controller will avoid syncing.", pod.Name)
+		return nil
+	}
+
+	if len(pdbs) > 1 {
+		msg := fmt.Sprintf("Pod %q/%q matches multiple PodDisruptionBudgets.  Chose %q arbitrarily.", pod.Namespace, pod.Name, pdbs[0].Name)
+		glog.Warning(msg)
+		dc.recorder.Event(pod, api.EventTypeWarning, "MultiplePodDisruptionBudgets", msg)
+	}
+	return &pdbs[0]
+}
+
+func (dc *DisruptionController) getPodsForPdb(pdb *policy.PodDisruptionBudget) ([]*api.Pod, error) {
+	sel, err := unversioned.LabelSelectorAsSelector(pdb.Spec.Selector)
+	if sel.Empty() {
+		return []*api.Pod{}, nil
+	}
+	if err != nil {
+		return []*api.Pod{}, err
+	}
+	podList, err := dc.podLister.Pods(pdb.Namespace).List(sel)
+	if err != nil {
+		return []*api.Pod{}, err
+	}
+	pods := []*api.Pod{}
+	for _, p := range podList.Items {
+		obj, err := api.Scheme.DeepCopy(p)
+		if err != nil {
+			return []*api.Pod{}, err
+		}
+		pod := obj.(api.Pod)
+		pods = append(pods, &pod)
+	}
+	return pods, nil
+}
+
+func (dc *DisruptionController) worker() {
+	for {
+		func() {
+			key, quit := dc.queue.Get()
+			if quit {
+				return
```

Comment:
This will never actually return the worker. Can you follow the pattern used for example in the [deployment controller](https://github.com/kubernetes/kubernetes/blob/afe4977c9badb90f8c883d76fac07ab5c26d13d9/pkg/controller/deployment/deployment_controller.go#L418)?